### PR TITLE
Chainspecs and cross-client,cross-chain tests

### PR DESCRIPTION
**Pull request by meowsbits**
:warning: Based on #65 to greatly simplify diff. Please review that first. 

:rabbit: Inclusively supersedes #61. 

---

### TL;DR

- Can read and write Parity JSON chainspecs 
  + https://github.com/etclabscore/multi-geth/tree/feat/x-chainspec-types-and-tests/params/parity.json.d

```sh
> make test-multigeth-chainspecs
```

- Can generate State and Difficulty tests for any supported chain config/fork
  + https://github.com/meowsbits/tests/commit/fd9819cf7edd94b65aaf742e281378b9b42f020f
  + https://github.com/meowsbits/tests/commit/7504868d88aedb877f6a6f43e987aea1ec00e86d

```sh 
> make tests-generate
 ```

- Can prove that multi-geth's Features-not-Forks implementation is test-equivalent to ethereum/go-ethereum's Foundation-only implementation

```sh
> make test-multigeth-features
```

> This is, to a large extent, a revisitation and rebuild of https://github.com/multi-geth/multi-geth/pull/103, which was built adhoc in preparation of the ETC Atlantis fork. With ETC Agharta around the corner, the need for cross-client compatible non-Foundation chain tests has risen again.

---

### Full Notes

:bookmark: Please see https://github.com/etclabscore/multi-geth/wiki/PR:-Chainspecs-and-cross-chain,cross-chain-test-generation.

### CODE REVIEW

**Review by tzdybal — APPROVED** [review]

**Comment by meowsbits** [comment]
Going to merge.

Since this is so big and architecturally significant, and touches several chain configs and consensus implementations, I think we should put the resulting `master` version thru sanity and manual "regression"  tests to make sure we don't see any obvious unexpected or missing behavior.
